docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,592 @@
+the following layers are proposed:
+- `application/config.image.inspektor-gadget.io+yaml`: [Configuration file](#configuration-file)
+- `application/ebpf.image.inspektor-gadget.io+binary`: Compiled eBPF object.
+
+### Configuration File
+
+The configuration file contains the following information for a gadget:
+- `name`: Human readable name of the gadget, like `trace_exec`, `profile_ebpf`, etc.
+- `description`: Short description about the gadget.
+- `documentation`: Long text describing how to use the gadget. This should describe the different
+  parameters supported and can contain some practical examples about how to use it.
+- `columns`: Information about the event that the gadget provides. Each element represents a columns attribute as defined in pkg/columns/columninfo.go.
+
+## Gadget Implementation
+
+A gadget is composed by eBPF programs and maps that are compiled into an eBPF object. According to
+the names of the maps and programs, and the types used, Inspektor Gadget is able to understand how
+to parse, enrich and print information from the gadget.
+
+Currently the built-in gadgets have a strong categorization, however in this new containerized
+gadgets approach that categorization is softer, each gadget defines how it behaves according to the
+maps / types they use and it's possible that the same gadget implements different behaviors, like
+tracer and topper, etc.
+
+### Maps
+
+#### `RingBuf` and `PerfEventArray` with `print_` prefix (a.k.a tracers)
+
+These maps indicate that the gadget produces a stream of events. Inspektor Gadget uses the BTF
+information to format and print the events.
+
+```c
+// taken from gadgets/trace_open.bpf.c
+struct event {
+	__u64 timestamp;
+	__u32 pid;
+	__u32 uid;
+	__u32 gid;
+	mnt_ns_id_t mntns_id;
+	int ret;
+	int flags;
+	__u16 mode;
+	__u8 comm[TASK_COMM_LEN];
+	__u8 fname[NAME_MAX];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__type(value, struct event);
+} print_events SEC(".maps");
+```
+
+#### `HashMap` with `stats_` prefix (a.k.a toppers)
+
+These maps are used to implement toppers, i.e. gadgets that print a list of elements sorted by
+specific parameters, like TCP connections by sent bytes.
+
+Inspektor Gadget automatically reads and sorts those maps according to the configuration provided by the user:
+- Interval: How often to print the map contents
+- Sort by: Field within the event type to sort by
+
+#### BPF Iterators (a.k.a snapshotters)
+
+Programs of type `iter/` are automatically loaded and attached by Inspektor Gadget, then they are
+triggered and its output is parsed according to the BTF information provided by it.
+
+TODO: It's not clear how to provide that BTF information as there is not any place in the iter
+program to define it. One possibility is to define a global constant and manually associated that to
+the program, by using its name for instance.
+
+```c
+const volatile event_t myiterator_output = {};
```

Comment:
I think it can work as done in https://github.com/inspektor-gadget/inspektor-gadget/pull/1866. 

It's still not clear how to handle other kind of iterators, like socket ones that should be executed in different network namespaces.